Refactors and workarounds so testing of collect_stats.py can be enabled in CI

shunghsiyu opened the pull request:
This pull requests contains a series of commits to refactor/workaround several blockers that prevented us from testing collect_stat.py in CI.
- Remove access to sysfs CPU device in collect_stats.py 
- Disable ASAN leak detection when testing collect_stats.py
- Give the libasan in LD_PRELOAD when testing collect_stats.py

shunghsiyu committed:
Remove access to sysfs CPU device in collect_stats.py

The container environment on which we run CI does not seem to have CPU entries
in sysfs, which cause the testing of collect_stats.py to fail. Remove those as a
temporarily measure.

For more information, see #2.

shunghsiyu committed:
Re-enable testing of collect_stats.py

dcermak commented:
For some reason the debug builds fail with clang & ASAN with:
```
==127==ASan runtime does not come first in initial library list; you should either link runtime to your application or manually preload it with LD_PRELOAD.
```
This could be related to the `LD_LIBRARY_PATH` hackery that I had to do in https://github.com/SUSE/phoebe/blob/0e679270fc66e99bf468098b49f3d08cc91bab88/tests/meson.build#L10 to get ASAN to work with clang.

shunghsiyu commented:
Thanks for the pointer, it seems more hackery is needed to get that part working :disappointed: . I'm working on that right now.

I'll push the code soon, and let's see if there is some way to make testing `collect_stats.py` with address sanitizer less painful. Or whether we prefer simply skipping `collect_stats.py` test instead (when address sanitizer is used).

shunghsiyu committed:
Use different environment variable holder for collect_static.py test

collect_static.py needs several environmental variables to workaround Python
import and keep ASAN happy. Those are not needed by phoebe, so we use another
environment variable holder to track those.

Those are:
- PYTHONPATH for finding _phoebe.abi3.so for import.
- LD_PRELOAD to make sure ASAN runtime is loaded first. Needed because the
  Python executable that comes with the system will not load ASAN runtime first,
  since it is not compile with -fsanitizer=address.
- ASAN_OPTIONS to ignore memory leaks in external code.

shunghsiyu committed:
Breakup integration test into smaller smaller pieces

This is a continuation of work done in #11. While building of _phoebe.abi3.so
can be toggled and dynamically determined, the integration test is not aware of
that; this works right now because testing of collect_stat.py is disabled right
now due to #2.

Break up what was run_tests.sh into setup.sh (majority of run_tests.sh goes
there) and two test() defined in /tests/meson.build, one for phoebe and another
for collect_static.py. Since meson runs the tests in parallel by default,
setup.sh is given an arbitrary high value with is_parallel == false to ensure it
is ran before anything else.

shunghsiyu committed:
Always expose the abi target in /scripts/meson.build

The changes introduced in 2e0aef0407 to build the shared library needed by
collect_stats.py only in the presence of python3-cffi module is not really
elegant. It introduce 'if' conditional, and most importantly, hides the abi
target when python3-cffi is not found; which make it difficult to specify it as
a test dependency (so testing of collect_stats.py can be skipped if it is not
built).

Remove the conditional and let prog_python become a disabler() when it is not
found. Disabler propagation will take care of the rest, which ensures that 1)
the abi target is always visible and 2) attempt to build abi_src and abi target
only when prog_python is found.

shunghsiyu commented:
Hmm...there seems to be problems with pulling container images from registry.opensuse.org (e.g. `registry.opensuse.org/home/dancermak/containers/phoebe`), and that's why the tests are failing.

`podman pull registry.opensuse.org/home/dancermak/containers/phoebe:latest` works for me locally though, so it might be a Azure/GitHub action problem.

shunghsiyu commented:
I'll add some comments and push again tomorrow to re-trigger GitHub action. (I don't see any button where I can ask GitHub to run the checks again)

mvarlese commented:
> I'll add some comments and push again tomorrow to re-trigger GitHub action. (I don't see any button where I can ask GitHub to run the checks again)

Please don't. I kicked off the re-running of jobs

mvarlese commented:
@shunghsiyu The job is still failing; this time though is not an infrastructure issue, the actual test seems to fail instead.

dcermak commented on scripts/meson.build:
```diff
@@ -1,16 +1,20 @@
+  modules : ['cffi'],
+  disabler : true,
+)
+collector_script = files(meson.current_source_dir() / 'collect_stats.py')
```
```suggestion
collector_script = files('collect_stats.py')
```
Using `meson.current_source_dir()` in `files()` is not required.

shunghsiyu replied:
Thanks!

shunghsiyu committed:
Add file target for collect_stats.py

shunghsiyu committed:
Refactor environment variable manipulation code

Use dictionary instead of array to improve readability.

shunghsiyu commented:
I forgot that UBSAN also required a runtime and that LD_LIBRARY_PATH need to be set when `CC=clang && b_sanitize=undefined`, that is now fixed.

mvarlese commented:
Thanks for the patch!
